### PR TITLE
Enable pt-dependent JER-SF and phi-dependent JEC application in PAT [106X backport]

### PR DESCRIPTION
**Pull request by ahinzmann**
#### PR description:

Enable pt-dependent jet resolution scale factors in PAT. This is needed to apply Autumn18_V7 JER-SF, released last week.
Enable phi-dependent jet energy scale in PAT. This will be needed for 2018 JEC to deal with the HEM issue.
This PR has no impact with the current global tags in any release. However, it will enable using a new global tag (currently in the makings) with pt-dependent JER-SFs and later with phi-dependent JECs.

A test config to apply resolution smearing from a db file has been added under /test.
Here is an example db-file containing pt-dependent JER-SFs:
https://github.com/cms-jet/JRDatabase/blob/master/SQLiteFiles/Autumn18_V7_MC.db

#### PR validation:

Test config runs after inclusion of this PR. It was checked that this does not change the memory consumption, compared to using the old code with Autumn18_V1_MC.db (no pt-dependence).

#### if this PR is a backport please specify the original PR:

Backport of #28096

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by fabiocos** [comment]
please test

File: PhysicsTools/PatUtils/interface/SmearedJetProducerT.h [hunk 1]
```diff
@@ -37,289 +37,288 @@
...
+        smearedJets->push_back(jet);
+
+        continue;
+      }
+
+      double jet_resolution = resolution.getResolution(
+          {{JME::Binning::JetPt, jet.pt()}, {JME::Binning::JetEta, jet.eta()}, {JME::Binning::Rho, *rho}});
+      double jer_sf = resolution_sf.getScaleFactor({{JME::Binning::JetPt, jet.pt()}, {JME::Binning::JetEta, jet.eta()}},
+                                                   m_systematic_variation);
+      if (m_debug) {
```
Comment (slava77) [inline PhysicsTools/PatUtils/interface/SmearedJetProducerT.h]: minor: the master version has an empty line above ```if (m_debug) {```

File: JetMETCorrections/Modules/plugins/JetResolutionDemo.cc [hunk 2]
```diff
@@ -23,192 +23,183 @@
...
+
+    // We do the same thing to access the scale factors
+    float sf = res_sf.getScaleFactor({{JME::Binning::JetPt, jet.pt()}, {JME::Binning::JetEta, jet.eta()}});
+
+    // Access up and down variation of the scale factor
+    float sf_up =
+        res_sf.getScaleFactor({{JME::Binning::JetPt, jet.pt()}, {JME::Binning::JetEta, jet.eta()}}, Variation::UP);
+    float sf_down =
+        res_sf.getScaleFactor({{JME::Binning::JetPt, jet.pt()}, {JME::Binning::JetEta, jet.eta()}}, Variation::DOWN);
+    if (m_debug) {
```
Comment (slava77) [inline JetMETCorrections/Modules/plugins/JetResolutionDemo.cc]: minor: the master version has an empty line above ```if (m_debug) {```

**Review by slava77 — COMMENTED** [review]
no need to follow up on these; I'm leaving the notes here for the record of a tiny inconsistency in the backport

**Comment by slava77** [comment]
backport of #28096 
this is more of a copy-paste from the master version though; it includes clang-format changes on top of #28096

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/28097	ce23314 
- code changes are in line with the PR description; added to support additional parameterization in the JEC/JER
- the backport is essentially correct (https://github.com/cms-sw/cmssw/pull/28097#pullrequestreview-298745749) 
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by santocch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**